karabo_data/geometry2.py

```diff
@@ -0,0 +1,326 @@
+from cfelpyutils.crystfel_utils import load_crystfel_geometry
+from copy import copy
+import numpy as np
+from scipy.ndimage import affine_transform
+import warnings
+
+def _crystfel_format_vec(vec):
+    """Convert an array of 3 numbers to CrystFEL format like "+1.0x -0.1y"
+    """
+    s = '{:+}x {:+}y'.format(*vec[:2])
+    if vec[2] != 0:
+        s += ' {:+}z'.format(vec[2])
+    return s
+
+class AGIPDGeometryFragment:
+    ss_pixels = 64
+    fs_pixels = 128
+
+    def __init__(self, corner_pos, ss_vec, fs_vec):
+        self.corner_pos = corner_pos
+        self.ss_vec = ss_vec
+        self.fs_vec = fs_vec
+
+    @classmethod
+    def from_panel_dict(cls, d):
+        corner_pos = np.array([d['cnx'], d['cny'], d['coffset']])
+        ss_vec = np.array([d['ssx'], d['ssy'], d['ssz']])
+        fs_vec = np.array([d['fsx'], d['fsy'], d['fsz']])
+        return cls(corner_pos, ss_vec, fs_vec)
+
+    def corners(self):
+        return np.stack([
+            self.corner_pos,
+            self.corner_pos + (self.fs_vec * self.fs_pixels),
+            self.corner_pos + (self.ss_vec * self.ss_pixels) + (self.fs_vec * self.fs_pixels),
+            self.corner_pos + (self.ss_vec * self.ss_pixels),
+        ])
+
+    def centre(self):
+        return self.corner_pos + (.5 * self.ss_vec * self.ss_pixels) \
+                               + (.5 * self.fs_vec * self.fs_pixels)
+
+    def to_crystfel_geom(self, p, a):
+        name = 'p{}a{}'.format(p, a)
+        c = self.corner_pos
+        return CRYSTFEL_PANEL_TEMPLATE.format(
+            name=name, p=p,
+            min_ss=(a * 64), max_ss=((a * 64) + 63),
+            ss_vec=_crystfel_format_vec(self.ss_vec),
+            fs_vec=_crystfel_format_vec(self.fs_vec),
+            corner_x=c[0], corner_y=c[1], coffset=c[2],
+        )
+
+
+class AGIPD_1MGeometry:
+    pixel_size = 2e-7
+    def __init__(self, modules):
+        self.modules = modules  # List of 16 lists of 8 fragments
+
+    @classmethod
+    def from_quad_positions(cls, quad_pos, asic_gap=2, panel_gap=29):
```

Comment:
How did you define `asic_gap` and `panel_gap`?

Reply:
People have told me that the pixels at the edges of the asics are twice as wide as the others. As there are two edges next to each other, I took that to mean there is effectively a 2 pixel gap between adjacent asics. This seems to work pretty well from my experiments with CrystFEL.

`panel_gap=29` is from trial and error working with CrystFEL.